scylla-cdc/src/stream_generations.rs

```diff
@@ -0,0 +1,540 @@
+        let (secs, nanos) = time
+            .to_timestamp()
+            .ok_or("Uuid is not a timeuuid")?
+            .to_unix();
+        let timestamp =
+            chrono::Duration::seconds(secs as i64) + chrono::Duration::nanoseconds(nanos as i64);
+
+        self.fetch_generation_by_timestamp(&timestamp).await
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_next_generation_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time > ?
+    ORDER BY time ASC
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    /// Given a generation returns the next generation.
+    /// If given generation is currently operating, returns None.
+    /// Propagates errors.
+    pub async fn fetch_next_generation(
+        &self,
+        generation: &Generation,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let mut query = Query::new(self.get_next_generation_query());
+        select_consistency(&self.session, &mut query).await?;
+
+        let result = self
+            .session
+            .query(query, (Timestamp(generation.0),))
+            .await?
+            .rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_stream_ids_by_time_query(&self) -> String {
+        format!(
+            r#"
+    SELECT streams
+    FROM {}
+    WHERE time = ?;
+    "#,
+            self.streams_table_name
+        )
+    }
+
+    /// Given a generation return identifiers of all streams of this generation.
+    pub async fn fetch_stream_ids(
+        &self,
+        generation: &Generation,
+    ) -> Result<Vec<StreamID>, Box<dyn Error>> {
+        let mut result_vec = vec![];
+
+        let mut query = Query::new(self.get_stream_ids_by_time_query());
+        select_consistency(&self.session, &mut query).await?;
+        query.set_page_size(DEFAULT_PAGE_SIZE);
+
+        let mut rows = self
+            .session
+            .query_iter(query, (Timestamp(generation.0),))
+            .await?
+            .into_typed::<(Vec<Vec<u8>>,)>();
+
+        while let Some(next_row) = rows.next().await {
+            let (ids,) = next_row?;
+            for id in ids {
+                result_vec.push(StreamID(id));
+            }
+        }
+        Ok(result_vec)
+    }
+
+    // Return single row containing generation.
+    fn return_single_row(row: Option<Vec<Row>>) -> Result<Option<Generation>, Box<dyn Error>> {
+        if let Some(row) = row {
+            if let Some(row) = row.into_typed::<(chrono::Duration,)>().next() {
+                let time = row?.0;
+                return Ok(Some(Generation(time)));
+            }
+        }
+
+        Ok(None)
+    }
+}
+
+// Returns current cluster size in case of a success.
+async fn get_cluster_size(session: &Session) -> Result<usize, Box<dyn Error>> {
```

Comment:
Doesn't the driver store the topology knowledge? Can't the cluster size be obtained directly from `session`?
(it's still good to ensure that we have sufficiently fresh topology information, but it should be also possible from `session` to explicitly refresh it)

Reply:
> Doesn't the driver store the topology knowledge? Can't the cluster size be obtained directly from `session`?

The driver internally keeps information about the topology, but does not expose it to the users. I agree it would be good if it was available to users.

> (it's still good to ensure that we have sufficiently fresh topology information, but it should be also possible from `session` to explicitly refresh it)

Topology refresh is quite an expensive operation because it reads various details about all the nodes in the cluster (e.g. their tokens, dcs) and creates a relatively heavyweight object to keep that information. Here, we are interested in the cluster size only, so I think it's OK if this function stays in its current form.